pkg/secretutils/secret_manager.go

```diff
@@ -0,0 +1,118 @@
+	return SecretManager{
+		log:       log.WithName("secret_manager"),
+		client:    cacheClient,
+		apiReader: apiReader,
+	}
+}
+
+// getSecret retrieves a Secret from the cache if it is available, and from the
+// k8s API if not.
+func (sm *SecretManager) getSecret(key types.NamespacedName) (secret *corev1.Secret, err error) {
+	secret = &corev1.Secret{}
+
+	// Look for secret in the filtered cache
+	err = sm.client.Get(context.TODO(), key, secret)
+	if err == nil {
+		return secret, nil
+	}
+	if !k8serrors.IsNotFound(err) {
+		return nil, err
+	}
+
+	// Secret not in cache; check API directly for unlabelled Secret
+	err = sm.apiReader.Get(context.TODO(), key, secret)
+	if err != nil {
+		return nil, err
+	}
+
+	return secret, nil
+}
+
+// claimSecret ensures that the Secret has a label that will ensure it is
+// present in the cache (and that we can watch for changes), and optionally
+// that it has a particular owner reference.
+func (sm *SecretManager) claimSecret(secret *corev1.Secret, owner metav1.Object, ownerIsController bool) error {
+	log := sm.log.WithValues("secret", secret.Name, "secretNamespace", secret.Namespace)
+	needsUpdate := false
+	if !metav1.HasLabel(secret.ObjectMeta, LabelEnvironmentName) {
+		log.Info("settting secret environment label")
+		metav1.SetMetaDataLabel(&secret.ObjectMeta, LabelEnvironmentName, LabelEnvironmentValue)
+		needsUpdate = true
+	}
+	if owner != nil {
+		if ownerIsController {
+			if !metav1.IsControlledBy(secret, owner) {
+				log.Info("setting secret controller reference")
+				if err := controllerutil.SetControllerReference(owner, secret, sm.client.Scheme()); err != nil {
+					return errors.Wrap(err, "failed to set secret controller reference")
+				}
+				needsUpdate = true
+			}
+		} else {
+			alreadyOwned := false
+			ownerUID := owner.GetUID()
+			for _, ref := range secret.GetOwnerReferences() {
+				if ref.UID == ownerUID {
+					alreadyOwned = true
+					break
+				}
+			}
+			if !alreadyOwned {
+				log.Info("setting secret owner reference")
+				if err := controllerutil.SetOwnerReference(owner, secret, sm.client.Scheme()); err != nil {
+					return errors.Wrap(err, "failed to set secret owner reference")
+				}
+				needsUpdate = true
+			}
+		}
+	}
+
+	if needsUpdate {
+		if err := sm.client.Update(context.TODO(), secret); err != nil {
+			return errors.Wrap(err, fmt.Sprintf("failed to update secret %s in namespace %s", secret.ObjectMeta.Name, secret.ObjectMeta.Namespace))
+		}
+	}
+
+	return nil
+}
+
+// ObtainSecret retrieves a Secret and ensures that it has a label that will
+// ensure it is present in the cache (and that we can watch for changes), and
+// optionally that it has a particular owner reference. The owner reference may
+// optionally be a controller reference.
+func (sm *SecretManager) ObtainSecret(key types.NamespacedName, owner metav1.Object, ownerIsController bool) (*corev1.Secret, error) {
```

Comment:
name bikeshedding: the feeling is that `obtain` doesn't communicate properly the fact that not only the object is going to be retrieved, but it's also to be owned if it's required. Maybe something like `AcquireSecret` could sound a little bit better.

Reply:
On a second thought, I realized that even `acquire` may not work well when `owner` is nil, because in that case no ownership action will be issued. For that, I'd guess if it could be better to split this method in a `ObtainSecret(key)` (just returns the secret and apply the environment label if required) and a `AcquireSecret(key, owner, ownerIsController)` that also sets the owner